Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\rightarrow$Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and collaborators below:

In [1]:
NAME = "Nesma Silman"
COLLABORATORS = ""

---

# Information Visualization I 
## School of Information, University of Michigan

## Week 3: 
- Perception / Cognition

## Assignment Overview
### This assignment's objectives include:

- Review, refect, and apply the concepts of the perception pipeline. Justify how different encodings impact the effectiveness of a visualization depending on the human perception process.

!["Drowing"](assets/preattentive_resized.png)

<p style="text-align: center;"> Preattentive Processing </p>

- Recreate visualizations and propose new and alternative visualizations using [Altair](https://altair-viz.github.io/) 

### The total score of this assignment will be 100 points consisting of:
- Case study reflection: America’s Favorite ‘Star Wars’ Movies (And Least Favorite Characters) (30 points)
- Altair programming exercise (70 points)

### Resources:
- Article by [FiveThirtyEight](https://fivethirtyeight.com) available  [online](https://fivethirtyeight.com/features/americas-favorite-star-wars-movies-and-least-favorite-characters/) (Hickey, 2014)  
- Datasets from FiveThirtyEight, we have downloaded a subset of this data in the folder [./assets](assets)
    - The original dataset can be found at [FiveThirtyEight Star Wars Survey](https://github.com/fivethirtyeight/data/tree/master/star-wars-survey)
    
    
### A note about autograding
Because there are many ways to generate the correct visualization, we will not be using the autograder for this assignment. Compare your result to the provided samples. Try to get as close to our provided solution as possible.

## Part 1. Perception and Cognition (30 points)
Read the article ["America’s Favorite ‘Star Wars’ Movies (And Least Favorite Characters),"](https://fivethirtyeight.com/features/americas-favorite-star-wars-movies-and-least-favorite-characters/) and answer the following questions:


### 1.1 List the different data types in the following visualizations and their encodings (10 points)
Look at the following visualizations. For each, list the variable, their type, and the encoding used (e.g., Weight, quantitative, color, ...)

!["Drowing"](assets/how_rate_resized.png)
!["Drowing"](assets/have_seen_resized.png)

### Vis # 1: How People Rate the "Star Wars" Movies, has 3 Variables

1- (Movie list chuncks (top third/middle third/bottom third), Ordinal, Color)

2- (Movie Name, Nominal, Position)

3- (How often was a movie rated in each chunk, Quantitative, Length)

### Vis # 2: Which "Star Wars" Movies Have You seen, has 2 Variables

1- (Movie Name, Nominal, Position)

2- (% of being seen by the 835 respondents, Quantitative, Length)

### 1.2 Propose an alternative encoding for the following visualization. Compare the visualizations based on perception. (10 points)
Either hand-draw or use an application to create a sketched solution. Upload an image and describe the differences between your solution and the FiveThirtyEight image in terms of perception (specifically for the task of comparing one movie to another).
!["Drowing"](assets/have_seen_resized.png)

                                             Which "Star Wars" Movies Have You seen?
                                             of 835 respondents who have seen any film
 !["Drowing"](assets/Vis1.png)

The differences between my solution and the FiveThirtyEight image in terms of perception (specifically for the task of comparing one movie to another):

The vis. in the article didn't sort the movies according to the % of seen, which makes it hard for the  reader to compare between different movies and catch the difference quickly. To make comparision easier, i changed the bars to be vertical and sorted them from largest to smallest (positional effect). I also used different shades of color to achieve the target of easier comparison as the original vis. used the same color which again makes it harder to compare different movies.

 ### 1.3 Propose an alternative encoding for the following visualization. Compare the visualizations based on perception. (10 points)
 Again, either-hand draw or use an application to create a sketched solution. Upload an image and describe the differences between your solution and the FiveThirtyEight image in terms of perception (specifically for the task of comparing one movie to another).
 !["Drowing"](assets/how_rate_resized.png)

                                         How People Rate the "Star Wars" Movies
                            How often each film was rated in the top, middle and bottom third
                                     (by 471 respondents who have seen all six films)
 !["Drowing"](assets/Vis2.PNG)

The differences between my solution and the FiveThirtyEight image in terms of perception (specifically for the task of comparing one movie to another):

I used different mark shapes to differentiate between top third/middle third/bottom third, and i tried to match the shape to the order, for example the top third is the best so i chose a triangle directed to the top. For the middle third, i tried to use a neutral shape so i selected a rectangle. For the bottom third, i used a triangle directed to the bottom to indicate that this is the least category.

I also used the same colors used by the original vis (Green - blue - red) respectively because i find it matching with the ordering of the 3 categories.

Moreover, i added the factor of size in encoding to facilitate the comparison/observation of different points for the reader.
I believe that using different marks shapes/size/color for each category rather than using only bars will make it easier for the reader to spot the difference between movies and compare between different movies easily, and also will facilitate observing the ratings within the same movie on the 3 levels.

Finally, i sorted the movies according to the top third values descendingly because i thought this might help the reader spot the best movies easily.

## Part 2. Altair programming exercise (70 points)
We have provided you with some code and parts of the article [America’s Favorite ‘Star Wars’ Movies (And Least Favorite Characters)](https://fivethirtyeight.com/features/americas-favorite-star-wars-movies-and-least-favorite-characters/). This article is based on the dataset:

1. [StarWars](data/StarWars.csv) Created by FiveThirtyEight based on a survey ran through SurveyMonkey Audience, surveying 1,186 respondents from June 3 to 6 2014. Available [online] (https://github.com/fivethirtyeight/data/tree/master/star-wars-survey)

To earn points for this assignment, you must:

- Recreate the visualizations in the article (replace the images in the article with a code cell that creates a visualization). We provide one example. Each visualization is worth 10 points (40 points/ 10 each x 4 total ).

    - _Partial credit can be granted for each visualization (up to 5 points) if you provide the grammar of graphics description of the visualization without a functional Altair implementation_


- Propose one alternative visualization for one of the article visualizations. Add a short paragraph describing why your visualization is more *effective* based on principles of perception/cognition. (15 points/ 10 points plot + 5 justification)


- Propose a new visualization to complement a part of the article. Add a short paragraph justifying your decisions in terms of Perception/Cognition processes.  (15 points/ 10 points plot + 5 justification)


In [2]:
import pandas as pd
import altair as alt
import numpy as np
import math

In [3]:
# enable correct rendering
alt.renderers.enable('default')

RendererRegistry.enable('default')

In [4]:
# uses intermediate json files to speed things up
alt.data_transformers.enable('json')

DataTransformerRegistry.enable('json')

In [5]:
sw = pd.read_csv('assets/StarWars.csv', encoding='latin1')

In [6]:
# Some format is needed for the survey dataframe, we provide the formatted dataset in a dataframe 
sw = sw.rename(columns={'Have you seen any of the 6 films in the Star Wars franchise?':'seen_any_movie',
                        'Do you consider yourself to be a fan of the Star Wars film franchise?': 'fan',
                        'Which of the following Star Wars films have you seen? Please select all that apply.' : 'seen_EI',
                        'Unnamed: 4' : 'seen_EII',
                        'Unnamed: 5' : 'seen_EIII',
                        'Unnamed: 6' : 'seen_EIV',
                        'Unnamed: 7' : 'seen_EV',
                        'Unnamed: 8' : 'seen_EVI',
                        'Please rank the Star Wars films in order of preference with 1 being your favorite film in the franchise and 6 being your least favorite film.' : 'rank_EI',
                        'Unnamed: 10' : 'rank_EII',
                        'Unnamed: 11' : 'rank_EIII',
                        'Unnamed: 12' : 'rank_EIV',
                        'Unnamed: 13' : 'rank_EV',
                        'Unnamed: 14' : 'rank_EVI',
                        'Please state whether you view the following characters favorably, unfavorably, or are unfamiliar with him/her.' : 'Han Solo',
                        'Unnamed: 16' : 'Luke Skywalker',
                        'Unnamed: 17' : 'Princess Leia Organa',
                        'Unnamed: 18' : 'Anakin Skywalker',
                        'Unnamed: 19' : 'Obi Wan Kenobi',
                        'Unnamed: 20' : 'Emperor Palpatine',
                        'Unnamed: 21' : 'Darth Vader',
                        'Unnamed: 22' : 'Lando Calrissian',
                        'Unnamed: 23' : 'Boba Fett',
                        'Unnamed: 24' : 'C-3P0',
                        'Unnamed: 25' : 'R2 D2',
                        'Unnamed: 26' : 'Jar Jar Binks',
                        'Unnamed: 27' : 'Padme Amidala',
                        'Unnamed: 28' : 'Yoda',
                       })
sw = sw.drop([0])

In [7]:
# take a peak to look at the data
sw.sample(5)

RespondentID seen_any_movie  fan  \
598  3.290129e+09             No  NaN   
752  3.289804e+09            Yes  Yes   
225  3.290947e+09            Yes   No   
772  3.289751e+09            Yes  Yes   
156  3.291233e+09            Yes  Yes   

                                      seen_EI  \
598                                       NaN   
752  Star Wars: Episode I  The Phantom Menace   
225  Star Wars: Episode I  The Phantom Menace   
772  Star Wars: Episode I  The Phantom Menace   
156  Star Wars: Episode I  The Phantom Menace   

                                        seen_EII  \
598                                          NaN   
752  Star Wars: Episode II  Attack of the Clones   
225  Star Wars: Episode II  Attack of the Clones   
772                                          NaN   
156  Star Wars: Episode II  Attack of the Clones   

                                       seen_EIII  \
598                                          NaN   
752  Star Wars: Episode III  Revenge of the Sith   
225  Star Wars: Episode III  Revenge of the Sith   
772                                          NaN   
156  Star Wars: Episode III  Revenge of the Sith   

                              seen_EIV  \
598                                NaN   
752  Star Wars: Episode IV  A New Hope   
225  Star Wars: Episode IV  A New Hope   
772  Star Wars: Episode IV  A New Hope   
156  Star Wars: Episode IV  A New Hope   

                                          seen_EV  \
598                                           NaN   
752  Star Wars: Episode V The Empire Strikes Back   
225  Star Wars: Episode V The Empire Strikes Back   
772  Star Wars: Episode V The Empire Strikes Back   
156  Star Wars: Episode V The Empire Strikes Back   

                                     seen_EVI rank_EI  ...  \
598                                       NaN     NaN  ...   
752  Star Wars: Episode VI Return of the Jedi       5  ...   
225  Star Wars: Episode VI Return of the Jedi       1  ...   
772  Star Wars: Episode VI Return of the Jedi       4  ...   
156  Star Wars: Episode VI Return of the Jedi       4  ...   

                   Yoda       Which character shot first?  \
598                 NaN                               NaN   
752      Very favorably                               Han   
225      Very favorably                            Greedo   
772  Somewhat favorably  I don't understand this question   
156      Very favorably                            Greedo   

    Are you familiar with the Expanded Universe?  \
598                                          NaN   
752                                           No   
225                                           No   
772                                           No   
156                                          Yes   

    Do you consider yourself to be a fan of the Expanded Universe?ξ  \
598                                                NaN                 
752                                                NaN                 
225                                                NaN                 
772                                                NaN                 
156                                                 No                 

    Do you consider yourself to be a fan of the Star Trek franchise?  Gender  \
598                                                 No                Female   
752                                                Yes                  Male   
225                                                Yes                Female   
772                                                 No                  Male   
156                                                Yes                Female   

       Age   Household Income                         Education  \
598  30-44       $0 - $24,999                   Graduate degree   
752  45-60  $25,000 - $49,999                   Graduate degree   
225  45-60  $25,000 - $49,999  Some college or Associate degree   
772   > 60                NaN 

# America’s Favorite ‘Star Wars’ Movies (And Least Favorite Characters)

_Original article available at [FiveThirtyEight](https://fivethirtyeight.com/features/americas-favorite-star-wars-movies-and-least-favorite-characters/)_

By [Walt Hickey](https://fivethirtyeight.com/contributors/walt-hickey/)

Filed under [Movies](https://fivethirtyeight.com/tag/movies/)

Get the data on [GitHub](https://github.com/fivethirtyeight/data/tree/master/star-wars-survey)

This week, I caught a sneak peek [of the X-Wing fighter](http://www.wired.com/2014/07/star-wars-episode-vii-x-wing/) from the new “Star Wars” films in production. The forthcoming movies — and the middling response to the most recent trilogy — provide a perfect excuse to examine some questions I’ve long wanted answers to: How many people are “Star Wars” fans? Does the rest of America realize that “The Empire Strikes Back” is clearly the best of the bunch? Which characters are most well-liked and most hated? And who shot first, Han Solo or Greedo?

We ran a poll through [SurveyMonkey Audience](https://www.surveymonkey.com/mp/audience/), surveying 1,186 respondents from June 3 to 6 (the [data](https://github.com/fivethirtyeight/data/tree/master/star-wars-survey) is available [on GitHub](https://github.com/fivethirtyeight/data)). Seventy-nine percent of those respondents said they had watched at least one of the “Star Wars” films. This question, incidentally, had a substantial difference by gender: 85 percent of men have seen at least one “Star Wars” film compared to 72 percent of women. Of people who have seen a film, men were also more likely to consider themselves a fan of the franchise: 72 percent of men compared to 60 percent of women.

We then asked respondents which of the films they had seen. With 835 people responding, here’s the probability that someone has seen a given “Star Wars” film given that they have seen any Star Wars film:

 !["Sol1"](assets/have_seen_resized.png)

In [8]:
# Sample visualization

# We're going to fix the labels a bit so will create a mapping to the full names
episodes = ['EI', 'EII', 'EIII', 'EIV', 'EV', 'EVI']
names = {
    'EI' : 'The Phantom Meanance', 'EII' : 'Attack of the clones', 'EIII' : 'Revenge of the Sith', 
    'EIV': 'A New Hope', 'EV': 'The Empire Strikes Back', 'EVI' : 'The Return of the Jedi'
}

# we're also going to use this order to sort, so names_l will now have our sort order
names_l = [names[ep] for ep in episodes]

print("sort order: ",names_l)

sort order:  ['The Phantom Meanance', 'Attack of the clones', 'Revenge of the Sith', 'A New Hope', 'The Empire Strikes Back', 'The Return of the Jedi']


In [9]:
# let's do some data pre-processing... sw (star wars) has everything

# We want to only use those people who have seen at least one movie, let's get the people, toss NAs
# and get the total count

# find people who have at least on of the columns (seen_*) not NaN
seen_at_least_one = sw.dropna(subset=['seen_' + ep for ep in episodes],how='all')
total = len(seen_at_least_one)

print("total who have seen at least one: ", total)

total who have seen at least one:  835


In [10]:
# for each movie, we're going to calculate the percents and generate a new data frame
percs = []

# loop over each column and calculate the number of people who have seen the movie
# specifically, filter out the people who are *NaN* for a specific episode (e.g., ep_EII), count them
# and divide by the percent
for seen_ep in ['seen_' + ep for ep in episodes]:
    perc = len(seen_at_least_one[~ pd.isna(seen_at_least_one[seen_ep])]) / total
    percs.append(perc)
    
# at this point percs is holding our percentages

# now we're going use a trick to make tuples--pairing names with percents--using "zip" and then make a dataframe
tuples = list(zip([names[ep] for ep in episodes],percs))
seen_per_df = pd.DataFrame(tuples, columns = ['Name', 'Percentage'])
seen_per_df

Name  Percentage
0     The Phantom Meanance    0.805988
1     Attack of the clones    0.683832
2      Revenge of the Sith    0.658683
3               A New Hope    0.726946
4  The Empire Strikes Back    0.907784
5   The Return of the Jedi    0.883832

In [11]:
# ok, time to make the chart... let's make a bar chart (use mark_bar)
bars = alt.Chart(seen_per_df).mark_bar(size=20).encode(
    # encode x as the percent, and hide the axis
    x=alt.X(
        'Percentage',
        axis=None),
    y=alt.Y(
        # encode y using the name, use the movie name to label the axis, sort using the names_l
        'Name:N',
         axis=alt.Axis(tickCount=5, title=''),
         # we give the sorting order to avoid alphabetical order
         sort=names_l
    )
)

# at this point we don't really have a great plot (it's missing the annotations, titles, etc.)
bars


alt.Chart(...)

In [12]:
# we're going to overlay the text with the percentages, so let's make another visualization
# that's just text labels

text = bars.mark_text(
    align='left',
    baseline='middle',
    dx=3  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    # we'll use the percentage as the text
    text=alt.Text('Percentage:Q',format='.0%')
)

# finally, we're going to combine the bars and the text and do some styling
seen_movies = (text + bars).configure_mark(
    # we don't love the blue
    color='#008fd5'
).configure_view(
    # we don't want a stroke around the bars
    strokeWidth=0
).configure_scale(
    # add some padding
    bandPaddingInner=0.2
).properties(
    # set the dimensions of the visualization
    width=500,
    height=180
).properties(
    # add a title
    title="Which 'Star Wars' Movies Have you Seen?"
)

seen_movies

# note that we are NOT formatting this in the Five Thirty Eight Style yet... we'll leave that to you to figure out

alt.LayerChart(...)

So we can see that “Star Wars: Episode V — The Empire Strikes Back” is the film seen by the most number of people, followed by “Star Wars: Episode VI — Return of the Jedi.” Appallingly, more people reported seeing “Star Wars: Episode I — The Phantom Menace” than the original “Star Wars” (renamed “Star Wars: Episode IV — A New Hope”).

So, which movie is the best? We asked the subset of 471 respondents who indicated they have seen every “Star Wars” film to rank them from best to worst. From that question, we calculated the share of respondents who rated each film as their favorite.

 !["Sol1"](assets/best_movie_article_resized.png)


 ** Homework note: Click [here](assets/best_movie.png) to see a version of this plot generated in Altair.

### 2.1 What's the best 'Star Wars' movie? Recreate the above image using altair (10 POINTS)

In [13]:
# Let's first check what columns we have
sw.columns

Index(['RespondentID', 'seen_any_movie', 'fan', 'seen_EI', 'seen_EII',
       'seen_EIII', 'seen_EIV', 'seen_EV', 'seen_EVI', 'rank_EI', 'rank_EII',
       'rank_EIII', 'rank_EIV', 'rank_EV', 'rank_EVI', 'Han Solo',
       'Luke Skywalker', 'Princess Leia Organa', 'Anakin Skywalker',
       'Obi Wan Kenobi', 'Emperor Palpatine', 'Darth Vader',
       'Lando Calrissian', 'Boba Fett', 'C-3P0', 'R2 D2', 'Jar Jar Binks',
       'Padme Amidala', 'Yoda', 'Which character shot first?',
       'Are you familiar with the Expanded Universe?',
       'Do you consider yourself to be a fan of the Expanded Universe?ξ',
       'Do you consider yourself to be a fan of the Star Trek franchise?',
       'Gender', 'Age', 'Household Income', 'Education',
       'Location (Census Region)'],
      dtype='object')

Now, we need to select the respondents who saw all 6 movies,

In [14]:
ppl=sw.dropna(subset=['seen_EI','seen_EII','seen_EIII', 'seen_EIV', 'seen_EV', 'seen_EVI'], how='any')
len(ppl)

471

Then we need to calculate the % of respondents who chose each movie as their favorite,

In [15]:
ranks=ppl[['rank_EI', 'rank_EII',
       'rank_EIII', 'rank_EIV', 'rank_EV', 'rank_EVI']]
ranks.columns=seen_per_df['Name']
ranks.head()

Name The Phantom Meanance Attack of the clones Revenge of the Sith A New Hope  \
1                       3                    2                   1          4   
4                       5                    6                   1          2   
5                       5                    4                   6          2   
6                       1                    4                   3          6   
7                       6                    5                   4          3   

Name The Empire Strikes Back The Return of the Jedi  
1                          5                      6  
4                          4                      3  
5                          1                      3  
6                          5                      2  
7                          1                      2

In [16]:
# While working with the data below, i discovered a nan value, so i will modify it here.

ranks[ranks.isnull().values]

Name The Phantom Meanance Attack of the clones Revenge of the Sith A New Hope  \
385                     1                    2                 NaN          4   

Name The Empire Strikes Back The Return of the Jedi  
385                        5                      3

I assume this respondent left it nan because he assumed we will conclude it will be 6 since he used other ranks. So i will modify the value to be 6.

In [17]:
ranks.loc['385','Revenge of the Sith']='6'

In [18]:
ranks.loc[385]

Name
The Phantom Meanance         1
Attack of the clones         2
Revenge of the Sith        NaN
A New Hope                   4
The Empire Strikes Back      5
The Return of the Jedi       3
Name: 385, dtype: object

In [19]:
# Let's change the data to the long format
lng_ranks=ranks.melt()
lng_ranks.head()

Name value
0  The Phantom Meanance     3
1  The Phantom Meanance     5
2  The Phantom Meanance     5
3  The Phantom Meanance     1
4  The Phantom Meanance     6

In [20]:
# Now let's select only records ranked as # 1

fav=lng_ranks[lng_ranks['value']=='1']
fav.head()

Name value
3   The Phantom Meanance     1
16  The Phantom Meanance     1
47  The Phantom Meanance     1
64  The Phantom Meanance     1
65  The Phantom Meanance     1

In [21]:
# Now, let's group by movie name and count the ones records,
ones=fav.groupby('Name').agg(len)
ones.reset_index(inplace=True)
ones

Name  value
0               A New Hope    128
1     Attack of the clones     18
2      Revenge of the Sith     27
3  The Empire Strikes Back    169
4     The Phantom Meanance     47
5   The Return of the Jedi     82

In [22]:
# Now i want to get the # of people choosing each movie as their favorite divided by overall # of people:
ones['Perc']=ones['value']/len(ppl)
ones

Name  value      Perc
0               A New Hope    128  0.271762
1     Attack of the clones     18  0.038217
2      Revenge of the Sith     27  0.057325
3  The Empire Strikes Back    169  0.358811
4     The Phantom Meanance     47  0.099788
5   The Return of the Jedi     82  0.174098

In [23]:
# Recreate this image using Altair
# try to match the "538 style" as best you can (hint: look at the altair lab at the start of the semester)
first = alt.Chart(ones).mark_bar(size=20).encode(
    x=alt.X(
        'Perc',
        axis=None),
    y=alt.Y(
        'Name:N',
         axis=alt.Axis(tickCount=5, title='', labelColor='#8a9499',tickSize=10,tickColor='#bbc1c4',labelPadding=5,domainColor='#bbc1c4'),
         sort=names_l
    )
)

txt = first.mark_text(
    align='left',
    baseline='middle',fontWeight='bold',
    dx=3  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    # we'll use the percentage as the text
    text=alt.Text('Perc:Q',format='.0%')
)

# finally, we're going to combine the bars and the text and do some styling
which_first = (txt + first).configure_mark(
    # we don't love the blue
    color='#34adea'
).configure_view(
    # we don't want a stroke around the bars
    strokeWidth=0
).configure_scale(
    # add some padding
    bandPaddingInner=0.2
).properties(
    # set the dimensions of the visualization
    width=500,
    height=180
).properties(
    # add a title
    title="What's the best 'Star Wars' Movie?"
).configure_title(anchor='start',fontSize=24, dy=-15)

which_first.properties(background='#f5f5f5')

#raise NotImplementedError()

alt.LayerChart(...)

## Make sure to *style* your visualization to match the original the best you can

We can also drill down and find out, generally, how people rate the films. Overall, fans broke into two camps: those who preferred the original three movies and those who preferred the three prequels. People who said “The Empire Strikes Back” was their favorite were also likely to rate “A New Hope” and “Return of the Jedi” higher as well. Those who rated “The Phantom Menace” as the best film were more likely to rate prequels higher.

This chart shows how often each film was rated in the top third (best or second-best), the middle third (third or fourth) or the bottom third (second-worst or worst). It’s a more nuanced take on the series:

 !["Sol1"](assets/how_rate_resized.png)

 ** Homework note: Click [here](assets/people_rate.png) to see a version of this plot generated in Altair.

### 2.2 How people rate the 'Star Wars' movie? Recreate the above image using altair (10 POINTS)

In [24]:
# Let's go back to our "lng_ranks" dataset. But this time, we will count how many time each movie was rated in each chunk.
# We have 3 chuncks. 1 & 2 is the first, 3 & 4 is the second, 5 & 6 is the third.

In [25]:
lng_ranks.head(7)

Name value
0  The Phantom Meanance     3
1  The Phantom Meanance     5
2  The Phantom Meanance     5
3  The Phantom Meanance     1
4  The Phantom Meanance     6
5  The Phantom Meanance     4
6  The Phantom Meanance     5

In [26]:
# Now let's manipulate the data so that we have one column for each rank,

all_rnks=lng_ranks.groupby(['Name','value']).agg(len).reset_index()
all_rnks.columns=['Name','value','count']
sums=all_rnks.pivot(index='Name', columns='value', values='count')
sums.reset_index(inplace=True)
sums

value                     Name    1    2    3    4    5    6
0                   A New Hope  128  107   86   60   30   60
1         Attack of the clones   18   47   52   84  186   84
2          Revenge of the Sith   27   34   68  121  104  117
3      The Empire Strikes Back  169  133   71   33   49   16
4         The Phantom Meanance   47   30   47  129   76  142
5       The Return of the Jedi   82  120  147   44   26   52

In [27]:
sums.sum()

value
Name    A New HopeAttack of the clonesRevenge of the S...
1                                                     471
2                                                     471
3                                                     471
4                                                     471
5                                                     471
6                                                     471
dtype: object

#### The first time i ran the above code, it returned the sum of # 6 was 470 which is less by 1 than others, so i went up to the 'ranks' dataset and modified the value.

In [28]:
lng_ranks['value'].unique()

array(['3', '5', '1', '6', '4', '2', nan], dtype=object)

In [29]:
# Now we want to add columns to calculate our chunks,

sums['Top third']=(sums['1']+sums['2'])/len(ppl)
sums['Middle third']=(sums['3']+sums['4'])/len(ppl)
sums['Bottom third']=(sums['5']+sums['6'])/len(ppl)
sums

value                     Name    1    2    3    4    5    6  Top third  \
0                   A New Hope  128  107   86   60   30   60   0.498938   
1         Attack of the clones   18   47   52   84  186   84   0.138004   
2          Revenge of the Sith   27   34   68  121  104  117   0.129512   
3      The Empire Strikes Back  169  133   71   33   49   16   0.641189   
4         The Phantom Meanance   47   30   47  129   76  142   0.163482   
5       The Return of the Jedi   82  120  147   44   26   52   0.428875   

value  Middle third  Bottom third  
0          0.309979      0.191083  
1          0.288747      0.573248  
2          0.401274      0.469214  
3          0.220807      0.138004  
4          0.373673      0.462845  
5          0.405520      0.165605

In [30]:
sums=sums[['Name','Top third','Middle third','Bottom third']]

sums

value                     Name  Top third  Middle third  Bottom third
0                   A New Hope   0.498938      0.309979      0.191083
1         Attack of the clones   0.138004      0.288747      0.573248
2          Revenge of the Sith   0.129512      0.401274      0.469214
3      The Empire Strikes Back   0.641189      0.220807      0.138004
4         The Phantom Meanance   0.163482      0.373673      0.462845
5       The Return of the Jedi   0.428875      0.405520      0.165605

In [31]:
second1 = alt.Chart(sums).mark_bar(size=20, color='#47a244').encode(
    x=alt.X(
        'Top third',
        axis=None),
    y=alt.Y(
        'Name:N',
         sort=names_l,axis=alt.Axis(tickCount=5, title='', labelColor='#8a9499',tickSize=10,tickColor='#bbc1c4',labelPadding=5,domainColor='#bbc1c4'),
         
    )
).properties(title={'text':'Top third', "fontSize":11, 'dy': -5,'fontWeight':'normal'})

text1 = second1.mark_text(
    align='left',
    baseline='middle',fontWeight='normal',color='#47a244',
    dx=3  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    # we'll use the percentage as the text
    text=alt.Text('Top third:Q',format='.0%')
)

# finally, we're going to combine the bars and the text and do some styling
final1 = (text1 + second1).properties(
    # set the dimensions of the visualization
    width=70,
    height=180
)

final1


alt.LayerChart(...)

In [32]:
second2 = alt.Chart(sums).mark_bar(size=20, color='#1798d9').encode(
    x=alt.X(
        'Middle third',
        axis=None),
    y=alt.Y(
        'Name:N',
         sort=names_l,axis=None
    )
).properties(
    title={'text':'Middle third', "fontSize":11, 'dy': -5,'fontWeight':'normal','dx':11})

text2 = second2.mark_text(
    align='left',
    baseline='middle',fontWeight='normal',color='#1798d9',
    dx=3  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    # we'll use the percentage as the text
    text=alt.Text('Middle third:Q',format='.0%')
)

# finally, we're going to combine the bars and the text and do some styling
final2 = (text2 + second2).properties(
    # set the dimensions of the visualization
    width=60,
    height=180
)

final2

alt.LayerChart(...)

In [33]:
second3 = alt.Chart(sums).mark_bar(size=20, color='#ff1a1a').encode(
    x=alt.X(
        'Bottom third',
        axis=None),
    y=alt.Y(
        'Name:N',
         sort=names_l,axis=None
    )
).properties(
    title={'text':'Bottom third', "fontSize":11, 'dy': -5,'fontWeight':'normal','dx':11})
    
text3 = second3.mark_text(
    align='left',
    baseline='middle',fontWeight='normal',color='#ff1a1a',
    dx=3  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    # we'll use the percentage as the text
    text=alt.Text('Bottom third:Q',format='.0%')
)

# finally, we're going to combine the bars and the text and do some styling
final3 = (text3 + second3).properties(
    # set the dimensions of the visualization
    width=65,
    height=180
)

final3

alt.LayerChart(...)

In [34]:
second_f=alt.hconcat(final1, final2, final3,spacing=40)
second_f.properties(background='#f5f5f5').properties(title="How People Rate the 'Star Wars' Movies"
).configure_title(fontSize=24, dy=-15,subtitleFontSize=5).configure_view(stroke=None)

alt.HConcatChart(...)

In [35]:
# Recreate this image using altair here (10 POINTS)

# YOUR CODE HERE
#raise NotImplementedError()

Finally, we took a boilerplate format used by political favorability polls — “Please state whether you view the following characters favorably, unfavorably, or are unfamiliar with him/her” — and asked respondents to rate characters in the series.

 !["Sol1"](assets/char_ranking_resized.png)


 ** Homework note. Here's an example solution generated in Altair:
 !["Sol1"](assets/people_rate_s.png)


### 2.3 Star Wars' Characters Favorability Ratings. Recreate the above image using altair (10 POINTS)

Let's first prepare the dataset,

In [36]:
fav=sw[sw.columns[15:29]]
fav.head()

Han Solo      Luke Skywalker Princess Leia Organa  \
1      Very favorably      Very favorably       Very favorably   
2                 NaN                 NaN                  NaN   
3  Somewhat favorably  Somewhat favorably   Somewhat favorably   
4      Very favorably      Very favorably       Very favorably   
5      Very favorably  Somewhat favorably   Somewhat favorably   

       Anakin Skywalker      Obi Wan Kenobi   Emperor Palpatine  \
1        Very favorably      Very favorably      Very favorably   
2                   NaN                 NaN                 NaN   
3    Somewhat favorably  Somewhat favorably    Unfamiliar (N/A)   
4        Very favorably      Very favorably  Somewhat favorably   
5  Somewhat unfavorably      Very favorably    Very unfavorably   

          Darth Vader                             Lando Calrissian  \
1      Very favorably                             Unfamiliar (N/A)   
2                 NaN                                          NaN   
3    Unfamiliar (N/A)                             Unfamiliar (N/A)   
4      Very favorably                           Somewhat favorably   
5  Somewhat favorably  Neither favorably nor unfavorably (neutral)   

              Boba Fett               C-3P0               R2 D2  \
1      Unfamiliar (N/A)      Very favorably      Very favorably   
2                   NaN                 NaN                 NaN   
3      Unfamiliar (N/A)    Unfamiliar (N/A)    Unfamiliar (N/A)   
4  Somewhat unfavorably      Very favorably      Very favorably   
5        Very favorably  Somewhat favorably  Somewhat favorably   

      Jar Jar Binks       Padme Amidala                Yoda  
1    Very favorably      Very favorably      Very favorably  
2               NaN                 NaN                 NaN  
3  Unfamiliar (N/A)    Unfamiliar (N/A)    Unfamiliar (N/A)  
4    Very favorably      Very favorably      Very favorably  
5  Very unfavorably  Somewhat favorably  Somewhat favorably

In [37]:
choices=list(fav['Han Solo'].unique())
choices

['Very favorably',
 nan,
 'Somewhat favorably',
 'Neither favorably nor unfavorably (neutral)',
 'Somewhat unfavorably',
 'Unfamiliar (N/A)',
 'Very unfavorably']

In [38]:
fav_wd=fav.melt()
fav_wd.dropna(inplace=True)
fav_wd.head(7)

variable               value
0  Han Solo      Very favorably
2  Han Solo  Somewhat favorably
3  Han Solo      Very favorably
4  Han Solo      Very favorably
5  Han Solo      Very favorably
6  Han Solo      Very favorably
7  Han Solo      Very favorably

In [39]:
all_fav=fav_wd.groupby(['variable','value']).agg(len).reset_index()
all_fav.columns=['Name','Vote','Count']
all_fav.head()

Name                                         Vote  Count
0  Anakin Skywalker  Neither favorably nor unfavorably (neutral)    135
1  Anakin Skywalker                           Somewhat favorably    269
2  Anakin Skywalker                         Somewhat unfavorably     83
3  Anakin Skywalker                             Unfamiliar (N/A)     52
4  Anakin Skywalker                               Very favorably    245

In [40]:
three=all_fav.pivot(index='Name', columns='Vote', values='Count').reset_index()
three

Vote                  Name  Neither favorably nor unfavorably (neutral)  \
0         Anakin Skywalker                                          135   
1                Boba Fett                                          248   
2                    C-3P0                                           79   
3              Darth Vader                                           84   
4        Emperor Palpatine                                          213   
5                 Han Solo                                           44   
6            Jar Jar Binks                                          164   
7         Lando Calrissian                                          236   
8           Luke Skywalker                                           38   
9           Obi Wan Kenobi                                           43   
10           Padme Amidala                                          207   
11    Princess Leia Organa                                           48   
12                   R2 D2                                           57   
13                    Yoda                                           51   

Vote  Somewhat favorably  Somewhat unfavorably  Unfamiliar (N/A)  \
0                    269                    83                52   
1                    153                    96               132   
2                    229                    23                15   
3                    171                   102                10   
4                    143                    68               156   
5                    151                     8                15   
6                    130                   102               109   
7                    223                    63               148   
8                    219                    13                 6   
9                    159                     8                17   
10                   183                    58               164   
11                   210                    12                 8   
12                   185                    10                10   
13                   144                     8                10   

Vote  Very favorably  Very unfavorably  
0                245                39  
1                138                45  
2                474                 7  
3                310               149  
4                110               124  
5                610                 1  
6                112               204  
7                142                 8  
8                552                 3  
9                591                 7  
10               168                34  
11               547                 6  
12               562                 6  
13               605                 8

In [41]:
three.rename({'Neither favorably nor unfavorably (neutral)':'Neutral','Unfamiliar (N/A)':'Unfamiliar'},inplace=True,axis=1)

In [42]:
three.columns[1:]

Index(['Neutral', 'Somewhat favorably', 'Somewhat unfavorably', 'Unfamiliar',
       'Very favorably', 'Very unfavorably'],
      dtype='object', name='Vote')

In [43]:
three['Total']=three[three.columns[1:]].sum(axis=1)
three

Vote                  Name  Neutral  Somewhat favorably  Somewhat unfavorably  \
0         Anakin Skywalker      135                 269                    83   
1                Boba Fett      248                 153                    96   
2                    C-3P0       79                 229                    23   
3              Darth Vader       84                 171                   102   
4        Emperor Palpatine      213                 143                    68   
5                 Han Solo       44                 151                     8   
6            Jar Jar Binks      164                 130                   102   
7         Lando Calrissian      236                 223                    63   
8           Luke Skywalker       38                 219                    13   
9           Obi Wan Kenobi       43                 159                     8   
10           Padme Amidala      207                 183                    58   
11    Princess Leia Organa       48                 210                    12   
12                   R2 D2       57                 185                    10   
13                    Yoda       51                 144                     8   

Vote  Unfamiliar  Very favorably  Very unfavorably  Total  
0             52             245                39    823  
1            132             138                45    812  
2             15             474                 7    827  
3             10             310               149    826  
4            156             110               124    814  
5             15             610                 1    829  
6            109             112               204    821  
7            148             142                 8    820  
8              6             552                 3    831  
9             17             591                 7    825  
10           164             168                34    814  
11             8             547                 6    831  
12            10             562                 6    830  
13            10             605                 8    826

In [44]:
three['Favorable']=(three['Very favorably']+three['Somewhat favorably'])/three['Total']
three['Unfavorable']=(three['Very unfavorably']+three['Somewhat unfavorably'])/three['Total']
three['Unfamiliar']=(three['Unfamiliar'])/three['Total']
three['Neutral']=(three['Neutral'])/three['Total']
three

Vote                  Name   Neutral  Somewhat favorably  \
0         Anakin Skywalker  0.164034                 269   
1                Boba Fett  0.305419                 153   
2                    C-3P0  0.095526                 229   
3              Darth Vader  0.101695                 171   
4        Emperor Palpatine  0.261671                 143   
5                 Han Solo  0.053076                 151   
6            Jar Jar Binks  0.199756                 130   
7         Lando Calrissian  0.287805                 223   
8           Luke Skywalker  0.045728                 219   
9           Obi Wan Kenobi  0.052121                 159   
10           Padme Amidala  0.254300                 183   
11    Princess Leia Organa  0.057762                 210   
12                   R2 D2  0.068675                 185   
13                    Yoda  0.061743                 144   

Vote  Somewhat unfavorably  Unfamiliar  Very favorably  Very unfavorably  \
0                       83    0.063183             245                39   
1                       96    0.162562             138                45   
2                       23    0.018138             474                 7   
3                      102    0.012107             310               149   
4                       68    0.191646             110               124   
5                        8    0.018094             610                 1   
6                      102    0.132765             112               204   
7                       63    0.180488             142                 8   
8                       13    0.007220             552                 3   
9                        8    0.020606             591                 7   
10                      58    0.201474             168                34   
11                      12    0.009627             547                 6   
12                      10    0.012048             562                 6   
13                       8    0.012107             605                 8   

Vote  Total  Favorable  Unfavorable  
0       823   0.624544     0.148238  
1       812   0.358374     0.173645  
2       827   0.850060     0.036276  
3       826   0.582324     0.303874  
4       814   0.310811     0.235872  
5       829   0.917973     0.010856  
6       821   0.294762     0.372716  
7       820   0.445122     0.086585  
8       831   0.927798     0.019254  
9       825   0.909091     0.018182  
10      814   0.431204     0.113022  
11      831   0.910951     0.021661  
12      830   0.900000     0.019277  
13      826   0.906780     0.019370

In [45]:
three=three[['Name','Favorable','Neutral','Unfavorable','Unfamiliar']]

three.sort_values(by='Favorable', ascending=False,inplace=True)

In [46]:
# Our criteria for sorting the names will be the values of "favorable" sorted in descending order.
chr_srt=list(three['Name'])
chr_srt

['Luke Skywalker',
 'Han Solo',
 'Princess Leia Organa',
 'Obi Wan Kenobi',
 'Yoda',
 'R2 D2',
 'C-3P0',
 'Anakin Skywalker',
 'Darth Vader',
 'Lando Calrissian',
 'Padme Amidala',
 'Boba Fett',
 'Emperor Palpatine',
 'Jar Jar Binks']

In [47]:
third1 = alt.Chart(three).mark_bar(size=20, color='#47a244').encode(
    x=alt.X(
        'Favorable',
        axis=None),
    y=alt.Y(
        'Name:N',
         sort=chr_srt,axis=alt.Axis(tickCount=5, title='', labelColor='#8a9499',tickSize=10,tickColor='#bbc1c4',labelPadding=5,domainColor='#bbc1c4'),
         
    )
).properties(title={'text':'Favorable', "fontSize":11, 'dy': -5,'fontWeight':'normal'})

text31 = third1.mark_text(
    align='left',
    baseline='middle',fontWeight='normal',color='#47a244',
    dx=3  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    # we'll use the percentage as the text
    text=alt.Text('Favorable:Q',format='.0%')
)

# finally, we're going to combine the bars and the text and do some styling
final31 = (text31 + third1).properties(
    # set the dimensions of the visualization
    width=80,
    height=300
)

final31


alt.LayerChart(...)

In [48]:
third2 = alt.Chart(three).mark_bar(size=20, color='#1798d9').encode(
    x=alt.X(
        'Neutral',
        axis=None),
    y=alt.Y(
        'Name:N',
         sort=chr_srt,axis=None
    )
).properties(
    title={'text':'Neutral', "fontSize":11, 'dy': -5,'fontWeight':'normal','dx':30})

text32 = third2.mark_text(
    align='left',
    baseline='middle',fontWeight='normal',color='#1798d9',
    dx=3  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    # we'll use the percentage as the text
    text=alt.Text('Neutral:Q',format='.0%')
)

# finally, we're going to combine the bars and the text and do some styling
final32 = (text32 + third2).properties(
    # set the dimensions of the visualization
    width=20,
    height=300
)

final32

alt.LayerChart(...)

In [49]:
third3 = alt.Chart(three).mark_bar(size=20, color='##ff1a1a').encode(
    x=alt.X(
        'Unfavorable',
        axis=None),
    y=alt.Y(
        'Name:N',
         sort=chr_srt,axis=None
    )
).properties(
    title={'text':'Unfavorable', "fontSize":11, 'dy': -5,'fontWeight':'normal','dx':30})

text33 = third3.mark_text(
    align='left',
    baseline='middle',fontWeight='normal',color='#ff1a1a',
    dx=3  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    # we'll use the percentage as the text
    text=alt.Text('Unfavorable:Q',format='.0%')
)

# finally, we're going to combine the bars and the text and do some styling
final33 = (text33 + third3).properties(
    # set the dimensions of the visualization
    width=20,
    height=300
)

final33

alt.LayerChart(...)

In [50]:
third4 = alt.Chart(three).mark_bar(size=20, color='#929b94').encode(
    x=alt.X(
        'Unfamiliar',
        axis=None),
    y=alt.Y(
        'Name:N',
         sort=chr_srt,axis=None
    )
).properties(
    title={'text':'Unfamiliar', "fontSize":11, 'dy': -5,'fontWeight':'normal','dx':30})

text34 = third4.mark_text(
    align='left',
    baseline='middle',fontWeight='normal',color='#929b94',
    dx=3  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    # we'll use the percentage as the text
    text=alt.Text('Unfamiliar:Q',format='.0%')
)

# finally, we're going to combine the bars and the text and do some styling
final34 = (text34 + third4).properties(
    # set the dimensions of the visualization
    width=20,
    height=300
)

final34

alt.LayerChart(...)

In [51]:
third_f=alt.hconcat(final31, final32, final33,final34,spacing=40)
third_f.properties(background='#f5f5f5').properties(title="'Star Wars' Characters Favorability Ratings"
).configure_title(fontSize=24, dy=-15,subtitleFontSize=5).configure_view(stroke=None)

alt.HConcatChart(...)

In [52]:
# Recreate this image using altair here (10 POINTS)


# YOUR CODE HERE
#raise NotImplementedError()

 You read that correctly. Jar Jar Binks has a lower favorability rating than the actual personification of evil in the galaxy.

 And for those of you who want to know the impact that [historical revisionism](http://en.wikipedia.org/wiki/Han_shot_first) can have on a society:

 !["Sol1"](assets/shot_first_article_resized.png)


 ** Homework note: Click [here](assets/shot_first.png) to see a version of this plot generated in Altair.

### 2.4 Who shot first? Recreate the above image using altair (10 POINTS)

In [53]:
# Let's prepare our data,
shot_all=sw['Which character shot first?'].dropna()
shot=sw['Which character shot first?'].value_counts()/len(shot_all)
sht=pd.DataFrame(shot)
sht['Name']=sht.index
sht

Which character shot first?  \
Han                                                  0.392512   
I don't understand this question                     0.369565   
Greedo                                               0.237923   

                                                              Name  
Han                                                            Han  
I don't understand this question  I don't understand this question  
Greedo                                                      Greedo

In [54]:
len(shot_all)

828

In [55]:
# I will use the same ordering used in the example vis,
sht_srt=['Han','Greedo',"I don't understand this question"]

In [56]:
fourth = alt.Chart(sht).mark_bar(size=20).encode(
    x=alt.X(
        'Which character shot first?',
        axis=None),
    y=alt.Y(
        'Name:N',
         axis=alt.Axis(tickCount=5, title='', labelColor='#8a9499',tickSize=10,tickColor='#bbc1c4',labelPadding=5,domainColor='#bbc1c4'),
         sort=sht_srt
    )
)

text4 = fourth.mark_text(
    align='left',
    baseline='middle',fontWeight='bold',
    dx=3  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    # we'll use the percentage as the text
    text=alt.Text('Which character shot first?:Q',format='.0%')
)

# finally, we're going to combine the bars and the text and do some styling
final4 = (text4 + fourth).configure_mark(
    # we don't love the blue
    color='#34adea'
).configure_view(
    # we don't want a stroke around the bars
    strokeWidth=0
).configure_scale(
    # add some padding
    bandPaddingInner=0.2
).properties(
    # set the dimensions of the visualization
    width=500,
    height=100
).properties(
    # add a title
    title="Who shot first?"
).configure_title(anchor='start',fontSize=24, dy=-15)

final4.properties(background='#f5f5f5')

#raise NotImplementedError()

alt.LayerChart(...)

In [57]:
# Recreate this image using altair here (10 POINTS)


# YOUR CODE HERE
#raise NotImplementedError()

### 2.5.1 Make your own (15 points/ 10 points plot + 5 justification)

Propose and code an alternative visualization for one of the visualizations *already in the article*. Add a short paragraph describing why your visualization is more (or less) *effective* based on principles of perception/cognition. 

If you feel your visualization is worse, that's ok! Just tell us why.

In [58]:
# YOUR CODE HERE
#raise NotImplementedError()

I'm going to redesign the "What's the best 'Star Wars' Movie?" visualization

In [59]:
ones.sort_values(by='Perc', ascending=False, inplace=True)

In [60]:
ones['nprcnt']=round(ones['Perc'],0).astype(str)
ones

Name  value      Perc nprcnt
3  The Empire Strikes Back    169  0.358811    0.0
0               A New Hope    128  0.271762    0.0
5   The Return of the Jedi     82  0.174098    0.0
4     The Phantom Meanance     47  0.099788    0.0
2      Revenge of the Sith     27  0.057325    0.0
1     Attack of the clones     18  0.038217    0.0

In [61]:
srt_nms=list(ones['Name'])

In [62]:
five1 = alt.Chart(ones).mark_point(filled=True, shape='diamond',color='#34adea').encode(
    x=alt.X(
        'Name:N',
        axis=alt.Axis(domainColor='#8a9499',tickColor='#bbc1c4',title='Movie Name',labelColor='#8a9499',labelAngle=315,
                      labelFontSize=12,titleFontSize=12, grid=True,gridColor='#c7cfff'),sort=srt_nms

    ),
    y=alt.Y('Perc',
         axis=alt.Axis(domainColor='#8a9499',tickCount=5, grid=False,title='% of Votes', labelColor='#8a9499',
                       labelPadding=5,labelFontSize=13,titleFontSize=12,format='%'),
    ),
    size=alt.Size('value',legend=None,scale=alt.Scale(range=[1000, 20000]))
)

text5 = five1.mark_text(align='left',
    baseline='middle',fontWeight='bold',
    dx=10  # Nudges text to right so it doesn't appear on top of the bar
).encode(size=alt.Size('Perc',legend=None,scale=alt.Scale(range=[15, 120])),
    # we'll use the percentage as the text
    text=alt.Text('Perc:N',format='.0%')
    
)

# finally, we're going to combine the bars and the text and do some styling
which_first2 = (text5 + five1).configure_scale(
    # add some padding
    bandPaddingInner=0.02
)

#which_first2
which_first2.properties(
    # set the dimensions of the visualization
    width=500,
    height=200,
    title="What's the best 'Star Wars' Movie?",background='#f5f5f5'
).configure_title(fontSize=24, dy=-15).configure_view(stroke=None)

alt.LayerChart(...)

For the original vis., i believe that setting the bars without sorting them makes it hard to catch the best movie or compare between different movies, so i used a scatter plot sorted from largest to smallest to make it more visible and easier to compare. I also used mark size to express the ranks from best to worst to help achieving the same effectiveness target.

One problem i faced after concatenating both the scatterplot and the text plot is that the variability in mark size decreased dramatically after i added the text chart which i couldn't manipulate. So the variability of marks size is not really expressive of the variance between percentages.

I also left the maximum value of y-axis at 40% although it might be preferred to make it 100% because i don't think keeping it at 40% has any negative impact on the clarity of the visualization and it will not be misleading. 

### 2.5.2 Make your own (15 points/ 10 points plot + 5 justification)
Propose and code a *new visualization* to complement a part of the article. Add a short paragraph justifying your decisions in terms of Perception/Cognition processes.

If you feel your visualization is worse, that's ok! Just tell us why.

In [63]:
# YOUR CODE HERE
#raise NotImplementedError()

For the article, i believe we can make a visualization to present the figures mentioned in the part talking about gender ratio among respondents.

In [64]:
# Here, we are interested in mainly 3 columns. has the respondent seen any movie, does he consider himself a fan and 
# the 'Gender' column

fiv=sw[['seen_any_movie', 'fan','Gender']]
fiv.head()

seen_any_movie  fan Gender
1            Yes  Yes   Male
2             No  NaN   Male
3            Yes   No   Male
4            Yes  Yes   Male
5            Yes  Yes   Male

In [65]:
len(fiv)

1186

In [66]:
#First, i want to select responses that has a gender value.
fiv.dropna(subset=['Gender'],inplace=True, how='any')
fiv.head()

seen_any_movie  fan Gender
1            Yes  Yes   Male
2             No  NaN   Male
3            Yes   No   Male
4            Yes  Yes   Male
5            Yes  Yes   Male

In [67]:
len(fiv)

1046

In [68]:
# Great. Now let's start by counting the number of respondents per gender,
cnts=fiv['Gender'].value_counts().reset_index()
cnts

index  Gender
0  Female     549
1    Male     497

In [69]:
ml=cnts[cnts['index']=='Male']['Gender'][1]
ml

497

In [70]:
fml=cnts[cnts['index']=='Female']['Gender'][0]
fml

549

In [71]:
totl=fml+ml
totl

1046

In [72]:
totl=fml+ml
totl

1046

In [73]:
# Now let's see the overall stats,
gndr_sn=fiv[['Gender','seen_any_movie']].groupby(['Gender','seen_any_movie']).agg(len).reset_index()
gndr_sn.columns=['Gender','seen_any_movie','count']
gndr_sn

Gender seen_any_movie  count
0  Female             No    152
1  Female            Yes    397
2    Male             No     74
3    Male            Yes    423

In [74]:
# Let's see the % of males who have seen a movie,

msn=gndr_sn[gndr_sn['Gender']=='Male']
m_s_prc=((msn[msn['seen_any_movie']=='Yes']['count'])/ml)[3]
m_s_prc

0.8511066398390342

In [75]:
# Now, let's see the % of females who have seen a movie,


fsn=gndr_sn[gndr_sn['Gender']=='Female']
f_s_prc=((fsn[fsn['seen_any_movie']=='Yes']['count'])/fml)[1]
f_s_prc

0.7231329690346083

In [76]:
# Now let's check respondents who saw a movie and see if they consider themselves fans or not,

In [77]:
allseen=fiv[fiv['seen_any_movie']=='Yes']
fns=allseen.groupby(['fan','Gender']).agg(len).reset_index()
fns

fan  Gender  seen_any_movie
0   No  Female             159
1   No    Male             120
2  Yes  Female             238
3  Yes    Male             303

In [78]:
fans=fns.pivot_table(index='Gender',values='seen_any_movie', columns='fan').reset_index()
fans['total']=fans.sum(axis=1)
fans

fan  Gender   No  Yes  total
0    Female  159  238    397
1      Male  120  303    423

In [79]:
# So let's calculate the % of each gender considering themselves as fans
mfans=fans.loc[1,'Yes']
ffans=fans.loc[0,'Yes']
mfansp=mfans/fans.loc[1,'total']
ffansp=ffans/fans.loc[0,'total']
ffansp

0.5994962216624685

In [80]:
seen_data=pd.DataFrame({'Gender':['Male','Female'],'Count':[ml,fml],'Percent':[m_s_prc,f_s_prc],
                       'Seen':[(msn[msn['seen_any_movie']=='Yes']['count'][3])
                            ,(fsn[fsn['seen_any_movie']=='Yes']['count'])[1]],'fans':[mfansp,ffansp],'fans_count':[mfans,ffans]})
seen_data

Gender  Count   Percent  Seen      fans  fans_count
0    Male    497  0.851107   423  0.716312         303
1  Female    549  0.723133   397  0.599496         238

In [81]:
bar = alt.Chart(seen_data).mark_bar(size=30,color='#1798d9').encode(
    y=alt.Y('Gender',axis=alt.Axis(titleFontSize=15,labelFontSize=12)),
    x=alt.X('Count',axis=alt.Axis(title='Count of Respondents',titleAlign='center',
                                 titleFontWeight='bold',titleFontSize=15))
#    legend=alt.Legend()
    # & Respondents Who saw a Movie
).properties(
    width=alt.Step(120)  # controls width of bar.
)


gtext1 = bar.mark_text(

    baseline='middle',fontWeight='normal',dx=15,color='#1798d9',fontSize=15,
  # Nudges text to right so it doesn't appear on top of the bar
).encode(text=alt.Text('Count:Q'))

tick = alt.Chart(seen_data).mark_tick(
    color='white',
    thickness=2,
    size=40 * 0.9,  # controls width of tick.
).encode(
    y='Gender',
    x=alt.X('Seen')
)

gtext2 = bar.mark_text(align='center',baseline='middle',dir='rtl',
fontWeight='normal',dx=-60,color='white',fontSize=15,
  # Nudges text to right so it doesn't appear on top of the bar
).encode(text=alt.Text('Percent:N',format='.0%'))


tick2 = alt.Chart(seen_data).mark_tick(
    color='black',
    thickness=2,
    size=40 * 0.9,  # controls width of tick.
).encode(
    y='Gender',
    x=alt.X('fans_count')
)

gtext3 = bar.mark_text(align='center',baseline='middle',dir='rtl',
fontWeight='bold',dx=-140,color='black',fontSize=15,
  # Nudges text to right so it doesn't appear on top of the bar
).encode(text=alt.Text('fans:N',format='.0%'))

final= (bar + tick + gtext1+gtext2+tick2 + gtext3).properties(background='#f5f5f5',width=300, height=100)
final.properties(title=["White mark: % of respondents who've seen a film" , 
                        "Black mark: % of respondents who've seen a film and consider themselves fans"]).configure_title(
    fontSize=14,color='red',dy=-10)

alt.LayerChart(...)

My decision to represent this part from the article was because i felt it is a bit confusing, because it mentioned many percentages which are subsets of each other without any visualizations. So i assumed the reader won't be able to form an understanding of what these figures really mean. So i decided to add this visualization to show the bigger picture (total count of respondents), and the subsets of both the % of respondents who reported they have seen at least 1 movie, and the % of respondents considering themselves as fans from people who have seen at least 1 movie.

I have some concerns on my visualization, i'll mention them below,

1- I tried to match the colors of 538 in my visualization (the colors in the favorability chart). However, colors were not clear when they overlapped. So i had to use white and black.

2- I wanted to add annotations beside the black and white vertical lines to explain what they indicate, but i found out that the annotation option does not exist in Altair, so i had to use the chart title to write my notes.

3- For the % written in white and black, the higher % text should be written so that it's x- value is more (white 85% should be located on x value higher than white 72%). However, i couldn't find the option to change it. Same goes for the black text.

4- My visualization might give wrong indication that for male respondents, males who consider themselves fans are 72% of the overall respondents count (497) which is incorrect. The truth is that the fans % is 72% of the 85% who have seen at least 1 movie. Same goes for females %. So i believe that visualizing a % as a part of a larger % can be misleading if this was not clearly explained.